DataStructures/DynamicStacks/Node.java

```diff
@@ -0,0 +1,33 @@
+public class Node {
+    private Node next;
+    private int payload;
+
+    public Node(int payload, Node nextNode){
+        setPayload(payload);
+        setNext(nextNode);
+    }
+
+    public Node getNext(){
+        return next;
+    }
+
+    public int getPayload(){
+        return payload;
+    }
+
+    public void setNext(Node next){
+        this.next = next;
+    }
+
+    public void setPayload(int payload){
+        this.payload = payload;
+    }
+
+    public Node withPayload(int payload){
+        return next;
```

Comment:
Aha! We get to these guys. The `with*` methods are used to build something known as a **fluent API**. The idea is to provide an API (in the old sense of the term: application programming interface) that allow to call many methods on the same instance on seemingly a single line of code.

See below example:

<table>
<tr><th>Normal API</th><th>Fluent API</th></tr>
<tr><td>

```java
Node nd = new Node();
nd.setPayload(5);
nd.setNext(otherNd);
```
</td><td>

```java
Node nd = new Node()
        .withPayload(5)
        .withNext(otherNd);
```
</td></tr>
</table>

How would you implement `withPayload()` and `withNext()` to work like the API described above?

Reply:
Made a couple of attempts at this, and I'm unsure how I could successfully implement. I have left my attempt in the code but commented it out.

Reply:
50% of it is correct. Remember that we're trying to execute these `with*` methods to **modify the same instance that was created**. Returning a reference to `next` is returning a Node, but it's not the same instance that was created, and in fact it won't work in a Node for which we don't have a next:

```java
// Let's say I want the first node only, so no next is available.
Node nd = new Node() // This is good, we're creating a new empty Node so we can call methods on it
        .withPayload(5); // but here, the method withPayload returns next...

System.out.println(nd.getPayload()); // So what would print here?
```
It would actually not print, it will trigger a `NullPointerException` because we're trying to call `getPayload()` on a variable whose value is `null` since the current version of `withPayload` is returning the reference to the next node, which is `null`. 

You're forgetting the keyword `this`. In Java, using the reserved word `this` within a class is the way to make a reference to the current object/instance and we can access properties and methods of it.  In other words, your fluent methods should return `this`, that way calling `withPayload` returns me the same instance that was created and I can call another methods on the same instance.

Question then, can I use the word `this` within a static method?

Reply:
I believe I have found the solution to this. As for the question, no we can't. If I understand correctly, since static methods don't need to have an instance, there could be no reference to "this".

Reply:
Correct!